src/gui/speedlimitdialog.cpp

```diff
@@ -50,57 +62,110 @@ SpeedLimitDialog::~SpeedLimitDialog()
     delete m_ui;
 }
 
-// -2: if cancel
-long SpeedLimitDialog::askSpeedLimit(QWidget *parent, bool *ok, const QString &title, const long defaultVal, const long maxVal)
+void SpeedLimitDialog::askNewSpeedLimits(QWidget *parent, bool *ok, const QString &title, SpeedLimits &speedLimits)
 {
     if (ok) *ok = false;
 
     SpeedLimitDialog dlg(parent);
     dlg.setWindowTitle(title);
-    dlg.setupDialog((maxVal / 1024.), (defaultVal / 1024.));
+    dlg.setupDialog(speedLimits);
 
     if (dlg.exec() == QDialog::Accepted) {
         if (ok) *ok = true;
 
-        const int val = dlg.getSpeedLimit();
-        if (val < 0)
-            return 0;
-        return (val * 1024);
+        if (dlg.getUploadSpeedLimit() < 0)
+            speedLimits.uploadLimit = 0;
+        else
+            speedLimits.uploadLimit = (dlg.getUploadSpeedLimit() * 1024);
+
+        if (dlg.getDownloadSpeedLimit() < 0)
+            speedLimits.downloadLimit = 0;
+        else
+            speedLimits.downloadLimit = (dlg.getDownloadSpeedLimit() * 1024);
     }
+}
+
+void SpeedLimitDialog::hideShowWarning()
+{
+    if (m_isGlobalLimits) return;
 
-    return -2;
+    if (((m_ui->spinUploadLimit->value() > m_globalUploadLimit) && (m_globalUploadLimit > 0))
+            || ((m_ui->spinDownloadLimit->value() > m_globalDownloadLimit) && (m_globalDownloadLimit > 0))) {
```

Comment:
assign the predicates to a `bool` or two, IMO easier to comprehend. or at least assign `m_ui->spinUploadLimit->value()` to a variable.

Reply:
I did both, i hope it's not overkill.